### DISCUSSION TITLE
nginx and ip_hash: examples/cluster-nginx uses always the same node

### DISCUSSION THREAD

**Discussion by mcelotti** [socketio/socket.io#3778] [Q&A] [answered]
Hello,
I'm running the "examples/cluster-nginx" project and I noticed that nginx uses always the same node.

The README says:
_This will start four Socket.IO nodes, behind a nginx proxy which will loadbalance the requests (using the IP of the client, see ip_hash)._

I'm testing the whole setup using multiple clients with different IP addresses but I always end up calling the same node.
I also tried replacing "ip_hash" with "hash $remote_addr" (according to https://serverfault.com/a/839175/404826) but nothing changed.

Any ideas?

Thanks

**Comment by darrachequesne** [comment, marked as answer]
That may be due to the fact the `ip_hash` only uses the first 3 octets of the client address (xxx.yyy.zzz.0).

From http://nginx.org/en/docs/http/ngx_http_upstream_module.html#ip_hash:

> The first three octets of the client IPv4 address, or the entire IPv6 address, are used as a hashing key.

I have updated the example here: https://github.com/socketio/socket.io/blob/master/examples/cluster-nginx/nginx/nginx.conf

You can test it with multiple `client` containers and the `hash $remote_addr consistent;` directive:

```
$ docker-compose up --scale=client=10 client
client_9         | 
client_9         | > socket.io-chat@0.0.0 start /usr/src/app
client_9         | > node index.js
client_9         | 
client_9         | connected
client_9         | connected to Ringo
client_6         | 
client_6         | > socket.io-chat@0.0.0 start /usr/src/app
client_6         | > node index.js
client_6         | 
client_6         | connected
client_6         | connected to Paul
client_10        | 
client_10        | > socket.io-chat@0.0.0 start /usr/src/app
client_10        | > node index.js
client_10        | 
client_10        | connected
client_10        | connected to Ringo
client_8         | 
client_8         | > socket.io-chat@0.0.0 start /usr/src/app
client_8         | > node index.js
client_8         | 
client_8         | connected
client_8         | connected to Paul
client_5         | 
client_5         | > socket.io-chat@0.0.0 start /usr/src/app
client_5         | > node index.js
client_5         | 
client_5         | connected
client_5         | connected to John
client_3         | 
client_3         | > socket.io-chat@0.0.0 start /usr/src/app
client_3         | > node index.js
client_3         | 
client_3         | connected
client_3         | connected to George
client_2         | 
client_2         | > socket.io-chat@0.0.0 start /usr/src/app
client_2         | > node index.js
client_2         | 
client_2         | connected
client_2         | connected to George
client_4         | 
client_4         | > socket.io-chat@0.0.0 start /usr/src/app
client_4         | > node index.js
client_4         | 
client_4         | connected
client_4         | connected to John
client_7         | 
client_7         | > socket.io-chat@0.0.0 start /usr/src/app
client_7         | > node index.js
client_7         | 
client_7         | connected
client_7         | connected to Paul
client_1         | 
client_1         | > socket.io-chat@0.0.0 start /usr/src/app
client_1         | > node index.js
client_1         | 
client_1         | connected
client_1         | connected to Ringo
client_1         | 
client_1         | > socket.io-chat@0.0.0 start /usr/src/app
client_1         | > node index.js
client_1         | 
client_1         | connected
client_1         | connected to Ringo
```

**Reply by mcelotti** [reply]
Thanks @darrachequesne, 
in my message with "multiple clients" I meant several browsers running on machines with different IP addresses, sorry for not pointing that out.
Anyway I tried again with the latest changes but I always end up connected to "George".
I managed to get what I'm looking for using your HAProxy example (cluster-haproxy) and that works since my browsers connect to different backends as expected.

Thanks
